pkg/hs/urbit-king/lib/Urbit/Arvo/Common.hs

```diff
@@ -167,8 +181,19 @@ newtype Port = Port { unPort :: Word16 }
   deriving newtype (Eq, Ord, Show, Enum, Real, Integral, Num, ToNoun, FromNoun)
 
 -- @if
-newtype Ipv4 = Ipv4 { unIpv4 :: Word32 }
-  deriving newtype (Eq, Ord, Enum, Real, Integral, Num, ToNoun, FromNoun)
+newtype Ipv4 = Ipv4 { unIpv4 :: N.HostAddress }
+  deriving newtype (Eq, Ord, Enum)
+
+instance Serialize Ipv4 where
+  get = Ipv4 <$> N.tupleToHostAddress
+    <$> ((,,,) <$> getWord8 <*> getWord8 <*> getWord8 <*> getWord8)
+  put (Ipv4 (N.hostAddressToTuple -> (a, b, c, d))) = for_ [a, b, c, d] putWord8
```

Comment:
Let me see if I understand this.  `<$>` is infix fmap with the function on the left and the value on the right, so the flow of control sort of goes right-to-left.  So the second line says "get 4 word8s and put them in a tuple".  I assume this happens "in order", so that if the original string was `0x7f000001`, then the tuple will be `[0x7f, 0x0, 0x0, 0x1]`.  Is that correct?

Then the first line converts this tuple into a `N.HostAddress` and then makes it of type `Ipv4`.

And then for `put`, we reverse `N.hostAddressToTuple` to get the word8s, which we put back in.  This seems to add them in order from "left to right", so I guess putWord8 adds to the least significant side.

So this looks to me like it leaves the everything in big-endian (i.e. network) order instead of reversing at all.  Is that correct?

Reply:
The key to note is that `hostAddressToTuple` calls `ntohl()` in its implementation, and `tupleToHostAddress` calls `htonl()` in its. I only know this 'cause I read the source ... I can't say I'm much of a fan of `Network.Socket` after this.

Reply:
Joe is right in pointing out that the "tuple conversion" functions in Network.Socket do "reversals" on little-endian machines. Philip's read of the applicative notation is right, in that it preserves order.

The put and get functions operate in monads which statefully produce / consume a ByteString in the order it would go on the network. So putWord8 0x01 >> putWord8 0x02 is the same as putWord16be 0x0102, where "be" means "big endian." Now <*> and <$> are applicative notation; <$> is fmap, and <*> basically lets you curry additional monadic arguments to your function. You'll note that if you just have fmap, and you have a binary function a -> b -> c, and an f a and an f b, you don't have any way of getting an f c; you need an extra capability for that.

The main problem with Network.Socket was that it made HostAddress a type alias, rather than an opaque newtype. We should never have had direct access to the underlying Word32, and that's how we got into this mess.

That, or the Word32 should have been in network order on all machines (unless that would mean calling htonl at each syscall, in which case the first seems to be the superor option).

Reply:
> So this looks to me like it leaves the everything in big-endian (i.e. network) order instead of reversing at all. Is that correct?

@philipcmonk you are right, and this code is wrong. An example:

The ip address for localhost is `.127.0.0.1`. In network byte order, that is 0x1000007f; in (little-endian) host byte order, that is 0x7f000001. As returned by `hostAddressToTuple`, it is `(127, 0, 0, 1)`. `Network.Socket.HostAddress` is in network byte order, the tuple calls convert to/from host byte order.

This line specifically is wrong:

```
put (Ipv4 (N.hostAddressToTuple -> (a, b, c, d))) = for_ [a, b, c, d] putWord8
```

In our example, that becomes `for_ [127, 0, 0, 1] putWord8`, which evaluates to `0x1000007f`. The correct result would be 0x7f000001.

---

> ... the Word32 should have been in network order on all machines ...

@pilfer-pandex it is.